columnflow/production/btag.py

```diff
@@ -0,0 +1,197 @@
+from columnflow.columnar_util import set_ak_column, flat_np_view, layout_ak_array
+
+np = maybe_import("numpy")
+ak = maybe_import("awkward")
+
+
+@producer(
+    uses={
+        "Jet.hadronFlavour", "Jet.eta", "Jet.pt", "Jet.btagDeepFlavB",
+    },
+    # produced columns are defined in the init function below
+)
+def btag_weight(
+    self: Producer,
+    events: ak.Array,
+    jet_mask: ak.Array | type(Ellipsis) = Ellipsis,
+    **kwargs,
+) -> ak.Array:
+    """
+    B-tag scale factor weight producer. Requires an external file in the config as (e.g.)
+
+    .. code-block:: python
+
+        "btag_sf_corr": ("/afs/cern.ch/user/m/mrieger/public/mirrors/jsonpog-integration-d0a522ea/POG/BTV/2017_UL/btagging.json.gz", "v1"),  # noqa
+
+    as well as an auxiliary entry in the config to refer to the b-tag correction set.
+
+    .. code-block:: python
+
+        cfg.x.btag_sf_correction_set = "deepJet_shape"
+
+    In addition, JEC uncertainty sources are propagated and weight columns are written if an
+    auxiliary config entry ``btag_sf_jec_sources`` exists.
+
+    Resources:
+
+       - https://twiki.cern.ch/twiki/bin/view/CMS/BTagShapeCalibration?rev=26
+       - https://indico.cern.ch/event/1096988/contributions/4615134/attachments/2346047/4000529/Nov21_btaggingSFjsons.pdf
+    """
+    if self.dataset_inst.is_data:
+        return events
+
+    # get the total number of jets in the chunk
+    n_jets_all = len(flat_np_view(events.Jet.pt, axis=1))
+
+    # get flat inputs, evaluated at jet_mask
+    flavor = flat_np_view(events.Jet.hadronFlavour[jet_mask], axis=1)
+    abs_eta = flat_np_view(abs(events.Jet.eta[jet_mask]), axis=1)
+    pt = flat_np_view(events.Jet.pt[jet_mask], axis=1)
+    b_discr = flat_np_view(events.Jet.btagDeepFlavB[jet_mask], axis=1)
+
+    # helper to create and store the weight
+    def add_weight(syst_name, syst_direction, column_name):
+        # define a mask that selects the correct flavor to assign to, depending on the systematic
+        flavor_mask = Ellipsis
+        if syst_name in ["hf", "lf", "hfstats1", "hfstats2", "lfstats1", "lfstats2"]:
+            flavor_mask = flavor != 4
+        elif syst_name in ["cferr1", "cferr2"]:
+            flavor_mask = flavor == 4
+
+        # get the flat scale factors
+        sf_flat = self.btag_sf_corrector.evaluate(
+            syst_name if syst_name == "central" else f"{syst_direction}_{syst_name}",
+            flavor[flavor_mask],
+            abs_eta[flavor_mask],
+            pt[flavor_mask],
+            b_discr[flavor_mask],
+        )
+
+        # insert them into an array of ones whose length corresponds to the total number of jets
+        sf_flat_all = np.ones(n_jets_all, dtype=np.float32)
+        if jet_mask is Ellipsis:
+            indices = flavor_mask
+        else:
+            indices = flat_np_view(jet_mask)
+            if flavor_mask is not Ellipsis:
+                indices = np.where(indices)[0][flavor_mask]
```

Comment:
why not just `indices = indices & flavor_mask` ?

Reply:
Then the resulting `indices` would have the same shape as before, just with fewer `True`'s, whereas above, there is an actual reduction happening to match the shape of the arrays it is applied to later on.

Reply:
OK, then I probably overlooked something. I'll maybe take another look into these lines of code after merging, but for now we can resolve this comment :+1: